tests/linearizability/model.go

```diff
@@ -187,24 +197,32 @@ func stepDelete(state EtcdState, request EtcdRequest, response EtcdResponse) (bo
 		state.FailedWrite = &request
 		return true, state
 	}
-	// revision should never decrease
-	if response.Revision < state.LastRevision {
-		return false, state
-	}
 	deleteSucceeded := response.Deleted != 0
 	keySet := state.Value != ""
 
-	// non-existent key cannot be deleted.
-	if deleteSucceeded != keySet && state.FailedWrite == nil {
-		return false, state
-	}
-	//if key was deleted, response revision should increase
-	if deleteSucceeded && (response.Revision != state.LastRevision+1 || !keySet) && (state.FailedWrite == nil || response.Revision < state.LastRevision+2) {
-		return false, state
-	}
-	//if key was not deleted, response revision should not change
-	if !deleteSucceeded && state.LastRevision != response.Revision && state.FailedWrite == nil {
-		return false, state
+	if state.FailedWrite == nil {
+		// non-existent key cannot be deleted.
+		if deleteSucceeded != keySet {
+			return false, state
+		}
+		// if key was deleted, response revision should increase by one
+		if deleteSucceeded && response.Revision != state.LastRevision+1 {
+			return false, state
+		}
+		// if key was not deleted, response revision should not change
+		if !deleteSucceeded && state.LastRevision != response.Revision {
+			return false, state
+		}
+	} else {
+		// if key was set, revision must increase as either key was deleted or failed request was persisted.
+		if keySet && response.Revision < state.LastRevision+1 {
+			return false, state
+		}
+
+		// if delete succeeded (+1 rev) without any key, means that fail request was persisted (+1 rev).
```

Comment:
Could you, please elaborate here more. What do you mean by persisting failRequest (in MVCC) ?

Reply:
From client perspective etcd is a black box. When clients get an error, etcd might never received the request as well already committed it to storage.

Reply:
I also didn't catch that part, how can the delete revision jump by up to two here?

Reply:
Yes. That's the exact concern. The requests either failed or succeeded... but not both (Schrödinger's cat), so if `+2` is justified, please provide more in-detail explanation.

Reply:
Please see https://github.com/etcd-io/etcd/pull/14902#issuecomment-1340041130, pls pay attention to the comment I added.